### PR TITLE
Leverage ext-curl when possible to achieve high speed downloads

### PR DESCRIPTION
**Pull request by nicolas-grekas**
This PR brings hirak/prestissimo-like download performance to Flex, but with greater compatibility with Composer's native behavior, thanks to a new extension point that is going to be part of Composer 1.7 (https://github.com/composer/composer/pull/6985).

Meanwhile, it already provides concurrent connections, and auto-enables the curl-based downloader conditionally.

In short, it should be totally seamless.

### CODE REVIEW

File: src/ParallelDownloader.php [hunk 1]
```diff
@@ -12,104 +12,94 @@
...
+
+    public function setNextOptions(array $options)
+    {
+        $this->nextOptions = parent::getOptions() !== $options ? $options : [];
+
+        return $this;
     }
 
-    protected function callbackGet($notificationCode, $severity, $message, $messageCode, $bytesTransferred, $bytesMax)
+    public function callbackGet($notificationCode, $severity, $message, $messageCode, $bytesTransferred, $bytesMax)
```
Comment (fabpot) [inline src/ParallelDownloader.php]: Should we tag it as being `@internal`?

**Comment by nicolas-grekas** [comment]
PR read, works GREAT!

**Comment by nicolas-grekas** [comment]
PR ready for another try. All good on my side. Now able to do concurrent downloads on more files than prestissimo. As such, prestissimo is now disabled if found.

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Pull request merged**